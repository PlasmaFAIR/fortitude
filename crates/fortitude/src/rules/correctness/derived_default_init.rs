use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Edit, Fix, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for uninitialised pointer variables inside derived types
///
/// ## Why is this bad?
/// Pointers inside derived types are undefined by default, and their
/// status cannot be tested by intrinsics such as `associated`. Pointer
/// variables should be initialised by either associating them with another
/// variable, or associating to `null()`.
///
///
/// ## Examples
/// For example, this derived type:
///
/// ```f90
/// type mytype
///     real :: val1
///     integer :: val2
///
///     real, pointer :: pReal1
///
///     integer, pointer :: pInt1 => null()
///     integer, pointer :: pI1
///     integer, pointer :: pI2 => null(), pI3
/// end mytype
/// ```
/// will have the pointers `pReal1`, `pI1`, and `pI3` uninitialised
/// whenever it is created. Instead, they should be initialised like:
/// ```f90
/// type mytype
///     real :: val1
///     integer :: val2
///
///     real, pointer :: pReal1
///
///     integer, pointer :: pInt1 => null()
///     integer, pointer :: pI1 => null()
///     integer, pointer :: pI2 => null(), pI3  => null()
/// end mytype
/// ```
///
/// ## References
/// - Metcalf, M., Reid, J. and Cohen, M., 2018, _Modern Fortran Explained:
///   Incorporating Fortran 2018_, Oxford University Press, Section 8.5.3/8.5.4.
/// - Clerman, N. Spector, W., 2012, _Modern Fortran: Style and Usage_, Cambridge
///   University Press, Rule 136, p. 189.
#[derive(ViolationMetadata)]
pub(crate) struct MissingDefaultPointerInitalisation {
    var: String,
}

impl Violation for MissingDefaultPointerInitalisation {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { var } = self;
        format!("pointer component '{var}' does not have a default initialiser")
    }

    fn fix_title(&self) -> Option<String> {
        Some("Associate to a known value, such as 'null()'".to_string())
    }
}

impl AstRule for MissingDefaultPointerInitalisation {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        // Only operate on derived types
        if node.parent()?.kind() != "derived_type_definition" {
            return None;
        }

        let mut node_cursor = node.walk();

        // Only check pointer variables
        if !node.named_children(&mut node_cursor).any(|node| {
            node.kind() == "type_qualifier"
                && node
                    .to_text(src.source_text())
                    .unwrap_or("")
                    .to_lowercase()
                    .starts_with("pointer")
        }) {
            return None;
        }

        let violations: Vec<Diagnostic> = node
            .named_descendants()
            .filter(|node| node.kind() == "identifier")
            .filter(|node| match node.parent() {
                None => false,
                Some(parent) => parent.kind() == "variable_declaration",
            })
            .map(|node| {
                let var_name = node.to_text(src.source_text()).unwrap_or("").to_string();

                let init_var = format!(" => null()");
                let start_pos = node.end_textsize();
                let fix = Fix::unsafe_edit(Edit::insertion(init_var, start_pos));

                Diagnostic::from_node(Self { var: var_name }, &node).with_fix(fix)
            })
            .collect();

        Some(violations)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["variable_declaration"]
    }
}
