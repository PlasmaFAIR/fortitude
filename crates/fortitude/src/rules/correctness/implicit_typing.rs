/// Defines rules that raise errors if implicit typing is in use.
use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

pub fn implicit_statement_is_none(node: &Node) -> bool {
    if let Some(child) = node.child(1) {
        return child.kind() == "none";
    }
    false
}

pub fn child_is_implicit_none(node: &Node) -> bool {
    if let Some(child) = node.child_with_name("implicit_statement") {
        return implicit_statement_is_none(&child);
    }
    false
}

/// ## What does it do?
/// Checks for missing `implicit none`
///
/// ## Why is this bad?
/// 'implicit none' should be used in all modules and programs, as implicit typing
/// reduces the readability of code and increases the chances of typing errors.
#[derive(ViolationMetadata)]
pub(crate) struct ImplicitTyping {
    entity: String,
}

impl Violation for ImplicitTyping {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { entity } = self;
        format!("{entity} missing 'implicit none'")
    }
}
impl AstRule for ImplicitTyping {
    fn check(_settings: &Settings, node: &Node, _src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if !child_is_implicit_none(node) {
            let entity = node.kind().to_string();
            let block_stmt = node.child(0)?;
            return some_vec![Diagnostic::from_node(Self { entity }, &block_stmt)];
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["module", "submodule", "program"]
    }
}

/// ## What it does
/// Checks for missing `implicit none` in interfaces
///
/// ## Why is this bad?
/// Interface functions and subroutines require 'implicit none', even if they are
/// inside a module that uses 'implicit none'.
#[derive(ViolationMetadata)]
pub(crate) struct InterfaceImplicitTyping {
    name: String,
}

impl Violation for InterfaceImplicitTyping {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("interface '{name}' missing 'implicit none'")
    }
}

impl AstRule for InterfaceImplicitTyping {
    fn check(_settings: &Settings, node: &Node, _src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let parent = node.parent()?;
        if parent.kind() == "interface" && !child_is_implicit_none(node) {
            let name = node.kind().to_string();
            let interface_stmt = node.child(0)?;
            return some_vec![Diagnostic::from_node(Self { name }, &interface_stmt)];
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["function", "subroutine"]
    }
}

/// ## What it does
/// Checks for unnecessary `implicit none` in module procedures
///
/// ## Why is this bad?
/// If a module has 'implicit none' set, it is not necessary to set it in contained
/// functions and subroutines (except when using interfaces).
#[derive(ViolationMetadata)]
pub(crate) struct SuperfluousImplicitNone {
    entity: String,
}

impl AlwaysFixableViolation for SuperfluousImplicitNone {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { entity } = self;
        format!("'implicit none' set on the enclosing {entity}")
    }

    fn fix_title(&self) -> String {
        "Remove unnecessary 'implicit none'".to_string()
    }
}

impl AstRule for SuperfluousImplicitNone {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if !implicit_statement_is_none(node) {
            return None;
        }
        let parent = node.parent()?;
        if matches!(parent.kind(), "function" | "subroutine") {
            for ancestor in parent.ancestors() {
                let kind = ancestor.kind();
                match kind {
                    "module" | "submodule" | "program" | "function" | "subroutine" => {
                        if !child_is_implicit_none(&ancestor) {
                            continue;
                        }
                        let entity = kind.to_string();
                        let fix = Fix::safe_edit(node.edit_delete(src));
                        return some_vec![
                            Diagnostic::from_node(Self { entity }, node).with_fix(fix)
                        ];
                    }
                    "interface" => {
                        break;
                    }
                    _ => {
                        continue;
                    }
                }
            }
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["implicit_statement"]
    }
}

/// ## What it does
/// Checks if `implicit none` is missing `external`
///
/// ## Why is this bad?
/// `implicit none` disables implicit types of variables but still allows
/// implicit interfaces for procedures. Fortran 2018 added the ability to also
/// forbid implicit interfaces through `implicit none (external)`, enabling the
/// compiler to check the number and type of arguments and return values.
///
/// `implicit none` is equivalent to `implicit none (type)`, so the full
/// statement should be `implicit none (type, external)`.
#[derive(ViolationMetadata)]
pub(crate) struct ImplicitExternalProcedures {}

impl Violation for ImplicitExternalProcedures {
    #[derive_message_formats]
    fn message(&self) -> String {
        "'implicit none' missing 'external'".to_string()
    }

    fn fix_title(&self) -> Option<String> {
        Some("Add `(external)` to 'implicit none'".to_string())
    }
}

impl AstRule for ImplicitExternalProcedures {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if !implicit_statement_is_none(node) {
            return None;
        }

        let text = node.to_text(src.source_text())?.to_lowercase();

        if !text.contains("external") {
            let edit = if let Some(type_node) = node
                .children(&mut node.walk())
                .find(|child| child.to_text(src.source_text()).unwrap().to_lowercase() == "type")
            {
                // Seems unlikely someone would have `implicit none (type)`
                // without `external` -- is that a sign they _explicitly_ don't
                // want it? That's probably still unwise though
                Edit::insertion(", external".to_string(), type_node.end_textsize())
            } else {
                Edit::insertion(" (type, external)".to_string(), node.end_textsize())
            };
            let fix = Fix::unsafe_edit(edit);

            some_vec!(Diagnostic::from_node(Self {}, node).with_fix(fix))
        } else {
            None
        }
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["implicit_statement"]
    }
}
