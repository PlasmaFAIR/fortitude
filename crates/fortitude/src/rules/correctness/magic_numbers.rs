use crate::ast::FortitudeNode;
use crate::rules::utilities::literal_as_io_unit;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use itertools::Itertools;
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for use of literals when specifying array sizes
///
/// ## Why is this bad?
/// Prefer named constants to literal integers when declaring arrays. This makes
/// it easier to find similarly sized arrays in the codebase, as well as ensuring
/// they are consistently sized when specified in different places. Named
/// parameters also make it easier for readers to understand your code.
///
/// The values `0, 1, 2, 3, 4` are ignored by default.
///
/// TODO: Add user settings
///
/// ## Examples
/// Instead of:
/// ```f90
/// integer, dimension(10) :: x, y
/// ```
/// prefer:
/// ```f90
/// integer, parameter :: NUM_SPLINE_POINTS = 10
/// integer, dimension(NUM_SPLINE_POINTS) :: x, y
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct MagicNumberInArraySize {
    value: i32,
}

impl Violation for MagicNumberInArraySize {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { value } = self;
        format!("Magic number in array size, consider replacing {value} with named `parameter`")
    }
}

const DEFAULT_ALLOWED_LITERALS: &[i32] = &[0, 1, 2, 3, 4];

impl AstRule for MagicNumberInArraySize {
    fn check(_settings: &Settings, node: &Node, source: &SourceFile) -> Option<Vec<Diagnostic>> {
        // We're either looking for `type, dimension(X) :: variable` or `type :: variable(X)`
        let size = if node.kind() == "type_qualifier" {
            if node.child(0)?.to_text(source.source_text())?.to_lowercase() != "dimension" {
                return None;
            }
            node.child_with_name("argument_list")?
        } else {
            // sized_declarator
            node.child_with_name("size")?
        };

        let violations: Vec<_> = size
            .named_children(&mut size.walk())
            .filter_map(|child| match child.kind() {
                // Need to return a vec here to match next arm
                "number_literal" => Some(vec![child]),
                // This is `X:Y`, pull out the lower and upper bound separately
                "extent_specifier" => Some(
                    child
                        .named_children(&mut child.walk())
                        .filter(|extent| extent.kind() == "number_literal")
                        .collect_vec(),
                ),
                _ => None,
            })
            .flatten()
            .filter_map(|literal| {
                let value = literal
                    .to_text(source.source_text())?
                    .parse::<i32>()
                    .unwrap();
                if DEFAULT_ALLOWED_LITERALS.contains(&value) {
                    None
                } else {
                    Some((literal, value))
                }
            })
            .map(|(child, value)| Diagnostic::from_node(Self { value }, &child))
            .collect();

        Some(violations)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["sized_declarator", "type_qualifier"]
    }
}

/// ## What it does
/// Checks for literal integers as units in IO statements.
///
/// ## Why is this bad?
/// Hardcoding unit numbers makes programs more brittle as it becomes harder to
/// verify units have been opened before reading/writing. Instead, units should
/// be passed in to procedures as arguments, or the `newunit=` argument used for
/// `open` statements. Having a named variable also makes it much clearer what a
/// given IO statement is for, and allows tools like LSP and IDEs to find all
/// references.
///
/// Bad:
/// ```f90
/// open(10, file="example.txt", action="read")
/// read(10, fmt=*) int
/// close(10)
/// ```
///
/// Good:
/// ```f90
/// open(newunit=example_unit, file="example.txt", action="read")
/// read(example_unit, fmt=*) int
/// close(example_unit)
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct MagicIoUnit {
    value: i32,
}

impl Violation for MagicIoUnit {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { value, .. } = self;
        format!("Magic unit '{value}' in IO statement")
    }

    fn fix_title(&self) -> Option<String> {
        Some("Replace with named variable".to_string())
    }
}

impl AstRule for MagicIoUnit {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let unit = literal_as_io_unit(node, src)?;

        let value = unit
            .to_text(src.source_text())?
            .parse::<i32>()
            .unwrap_or_default();

        some_vec!(Diagnostic::from_node(Self { value }, &unit))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec![
            "read_statement",
            "write_statement",
            "open_statement",
            "close_statement",
        ]
    }
}
