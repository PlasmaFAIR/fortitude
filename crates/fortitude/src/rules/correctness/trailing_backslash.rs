use crate::ast::FortitudeNode;
/// Defines rules that govern line length.
use crate::settings::Settings;
use crate::AstRule;
use lazy_regex::regex;
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use ruff_text_size::{TextRange, TextSize};
use tree_sitter::Node;

/// ## What does it do?
/// Checks if a backslash is the last character on a line
///
/// ## Why is this bad?
/// When compilers use the C preprocessor to pre-process Fortran files
/// the \ character is treated as a line continuation character by the C preprocessor,
/// potentially causing lines to be merged into one.
///
/// ## Example
/// When this Fortran program is passed through the C preprocessor,
/// ```f90
/// program t
///     implicit none
///     real :: A
///
///     ! Just a comment \
///     A = 2.0
///     print *, A
///  end
/// ```
/// it will end up with the variable assignment A placed onto the comment line,
/// ```f90
/// program t
///    implicit none
///    real :: A
///
///    ! Just a comment    A = 2.0
///
///    print *, A
/// end
/// ```
/// which causes the assignment to not be compiled.
///
#[derive(ViolationMetadata)]
pub(crate) struct TrailingBackslash {}

impl Violation for TrailingBackslash {
    #[derive_message_formats]
    fn message(&self) -> String {
        "Trailing backslash".to_string()
    }
}

impl AstRule for TrailingBackslash {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        // Preprocessor might ignore trailing whitespace
        let trailing_backslash_re = regex!(r#".*(\\)\s*$"#);

        let comment = node.to_text(src.source_text())?;
        let captures = trailing_backslash_re.captures(comment)?;

        let trailing_backslash = captures.get(1)?;
        let start: TextSize = trailing_backslash.start().try_into().unwrap();
        let end: TextSize = trailing_backslash.end().try_into().unwrap();
        // Regex start/end are relative to start of comment node
        let comment_start: TextSize = node.start_byte().try_into().unwrap();
        let range = TextRange::new(comment_start + start, comment_start + end);
        some_vec!(Diagnostic::new(Self {}, range))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["comment"]
    }
}
