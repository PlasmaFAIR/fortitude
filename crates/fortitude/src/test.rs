use std::path::Path;

use anyhow::Result;
use ruff_source_file::{SourceFile, SourceFileBuilder};

use crate::{
    check::{
        ast_entrypoint_map, check_file, read_to_string, rules_to_path_rules, rules_to_text_rules,
    },
    message::{Emitter, TextEmitter},
    rule_table::RuleTable,
    rules::Rule,
    settings::{self, FixMode, Settings},
};

#[macro_export]
macro_rules! apply_common_filters {
    {} => {
        let mut settings = insta::Settings::clone_current();
        // Convert windows paths to Unix Paths.
        settings.add_filter(r"\\\\?([\w\d.])", "/$1");
        let _bound = settings.bind_to_scope();
    }
}

pub(crate) fn test_resource_path(path: impl AsRef<Path>) -> std::path::PathBuf {
    Path::new("./resources/test/").join(path)
}

/// Run [`check_path`] on a Fortran file in the `resources/test/fixtures` directory.
pub(crate) fn test_path(
    path: impl AsRef<Path>,
    rules: &[Rule],
    settings: &Settings,
) -> Result<String> {
    let path = test_resource_path("fixtures").join(path);
    let source = read_to_string(&path)?;
    let filename = path.to_string_lossy();
    let file = SourceFileBuilder::new(filename.as_ref(), source.as_str()).finish();
    Ok(test_contents(&path, &file, rules, settings))
}

pub(crate) fn test_contents(
    path: &Path,
    file: &SourceFile,
    rules: &[Rule],
    settings: &Settings,
) -> String {
    let rule_table = RuleTable::from_iter(rules.iter().cloned());
    let path_rules = rules_to_path_rules(&rule_table);
    let text_rules = rules_to_text_rules(&rule_table);
    let ast_entrypoints = ast_entrypoint_map(&rule_table);

    match check_file(
        &rule_table,
        &path_rules,
        &text_rules,
        &ast_entrypoints,
        path,
        file,
        settings,
        FixMode::Generate,
        settings::IgnoreAllowComments::Disabled,
    ) {
        Ok(violations) => {
            if violations.messages.is_empty() {
                return String::new();
            }

            let mut output = Vec::new();

            TextEmitter::default()
                .with_show_fix_status(true)
                .with_show_fix_diff(true)
                .with_show_source(true)
                .with_unsafe_fixes(crate::settings::UnsafeFixes::Enabled)
                .emit(&mut output, &violations.messages)
                .unwrap();

            String::from_utf8(output).unwrap()
        }
        Err(msg) => {
            panic!("Failed to process: {msg}");
        }
    }
}
