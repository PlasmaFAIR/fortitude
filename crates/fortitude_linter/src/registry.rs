use fortitude_macros::RuleNamespace;
use rayon::str;
use std::str::FromStr; // Needed by strum_macros

pub use crate::rules::Rule;
pub use rule_set::{RuleSet, RuleSetIterator};

mod rule_set;

// Rule categories and identity codes
// ----------------------------------
// Helps to sort rules into logical categories, and acts as a unique tag with which
// users can switch rules on and off.

pub trait AsRule {
    fn rule(&self) -> Rule;
}

impl Rule {
    pub fn from_code(code: &str) -> Result<Self, FromCodeError> {
        let (category, code) = Category::parse_code(code).ok_or(FromCodeError::Unknown)?;
        category
            .all_rules()
            .find(|rule| rule.noqa_code().suffix() == code)
            .ok_or(FromCodeError::Unknown)
    }
}

#[derive(thiserror::Error, Debug)]
pub enum FromCodeError {
    #[error("unknown rule code")]
    Unknown,
}

/// The category of each rule defines the sort of problem it intends to solve.
#[derive(
    Debug,
    PartialEq,
    Eq,
    Clone,
    Hash,
    PartialOrd,
    Ord,
    strum_macros::AsRefStr,
    strum_macros::Display,
    strum_macros::EnumIter,
    strum_macros::EnumString,
    strum_macros::IntoStaticStr,
    RuleNamespace,
)]
#[repr(u16)]
#[strum(serialize_all = "kebab-case")]
pub enum Category {
    /// Failure to parse a file.
    #[prefix = "E"]
    Error,
    /// Detect code that is bug-prone or likely to be incorrect.
    #[prefix = "C"]
    Correctness,
    /// Obsolescent language features, as determined by the Fortran standard.
    #[prefix = "OB"]
    Obsolescent,
    /// Update to modern Fortran features. Used for less severe issues than `Obsolescent`,
    /// and goes beyond recommendations in the Fortran standard.
    #[prefix = "MOD"]
    Modernisation,
    /// Violation of style conventions.
    #[prefix = "S"]
    Style,
    /// Avoid platform/compiler-specific features.
    #[prefix = "PORT"]
    Portability,
    /// Fortitude specific rules.
    #[prefix = "FORT"]
    Fortitude,
}

pub trait RuleNamespace: Sized {
    /// Returns the prefix that every single code that fortitude uses to identify
    /// rules from this category starts with.
    fn common_prefix(&self) -> &'static str;

    /// Returns the variant corresponding to the given prefix.
    fn from_prefix(preifx: &str) -> Result<Self, &'static str>;

    /// Attempts to parse the given rule code. If the prefix is recognized
    /// returns the respective variant along with the code with the common
    /// prefix stripped.
    fn parse_code(code: &str) -> Option<(Self, &str)>;

    #[allow(dead_code)]
    fn name(&self) -> &'static str;

    #[allow(dead_code)]
    fn description(&self) -> &'static str;
}

pub mod clap_completion {
    use clap::builder::{PossibleValue, TypedValueParser, ValueParserFactory};
    use strum::IntoEnumIterator;

    use crate::registry::Rule;

    #[derive(Clone)]
    pub struct RuleParser;

    impl ValueParserFactory for Rule {
        type Parser = RuleParser;

        fn value_parser() -> Self::Parser {
            RuleParser
        }
    }

    impl TypedValueParser for RuleParser {
        type Value = Rule;

        fn parse_ref(
            &self,
            cmd: &clap::Command,
            arg: Option<&clap::Arg>,
            value: &std::ffi::OsStr,
        ) -> Result<Self::Value, clap::Error> {
            let value = value
                .to_str()
                .ok_or_else(|| clap::Error::new(clap::error::ErrorKind::InvalidUtf8))?;

            Rule::from_code(value).map_err(|_| {
                let mut error =
                    clap::Error::new(clap::error::ErrorKind::ValueValidation).with_cmd(cmd);
                if let Some(arg) = arg {
                    error.insert(
                        clap::error::ContextKind::InvalidArg,
                        clap::error::ContextValue::String(arg.to_string()),
                    );
                }
                error.insert(
                    clap::error::ContextKind::InvalidValue,
                    clap::error::ContextValue::String(value.to_string()),
                );
                error
            })
        }

        fn possible_values(&self) -> Option<Box<dyn Iterator<Item = PossibleValue> + '_>> {
            Some(Box::new(Rule::iter().map(|rule| {
                let name = rule.noqa_code().to_string();
                let help = rule.as_ref().to_string();
                PossibleValue::new(name).help(help)
            })))
        }
    }
}

#[cfg(test)]
mod tests {
    use std::mem::size_of;

    use strum::IntoEnumIterator;

    use super::{Category, Rule, RuleNamespace};

    #[test]
    fn documentation() {
        for rule in Rule::iter() {
            assert!(
                rule.explanation().is_some(),
                "Rule {} is missing documentation",
                rule.as_ref()
            );
        }
    }

    #[test]
    fn check_code_serialization() {
        for rule in Rule::iter() {
            assert!(
                Rule::from_code(&format!("{}", rule.noqa_code())).is_ok(),
                "{rule:?} could not be round-trip serialized."
            );
        }
    }

    #[test]
    fn category_parse_code() {
        for rule in Rule::iter() {
            let code = format!("{}", rule.noqa_code());
            let (category, rest) =
                Category::parse_code(&code).unwrap_or_else(|| panic!("couldn't parse {code:?}"));
            assert_eq!(code, format!("{}{rest}", category.common_prefix()));
        }
    }

    #[test]
    fn rule_size() {
        assert_eq!(2, size_of::<Rule>());
    }
}
