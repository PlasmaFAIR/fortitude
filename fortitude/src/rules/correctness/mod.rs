pub mod accessibility_statements;
pub mod assumed_size;
pub mod derived_default_init;
pub mod exit_labels;
pub mod external;
pub mod implicit_kinds;
pub mod implicit_typing;
pub mod init_decls;
pub mod intent;
pub mod kind_suffixes;
pub mod magic_numbers;
pub mod missing_io_specifier;
pub mod select_default;
pub mod trailing_backslash;
pub mod use_statements;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::rules::correctness::exit_labels;
    use crate::settings::{CheckSettings, Settings};
    use crate::test::test_path;

    #[test_case(Rule::ImplicitTyping, Path::new("C001.f90"))]
    #[test_case(Rule::InterfaceImplicitTyping, Path::new("C002.f90"))]
    #[test_case(Rule::ImplicitExternalProcedures, Path::new("C003.f90"))]
    #[test_case(Rule::MissingDefaultCase, Path::new("C011.f90"))]
    #[test_case(Rule::NoRealSuffix, Path::new("C021.f90"))]
    #[test_case(Rule::ImplicitRealKind, Path::new("C022.f90"))]
    #[test_case(Rule::MagicNumberInArraySize, Path::new("C031.f90"))]
    #[test_case(Rule::MagicIoUnit, Path::new("C032.f90"))]
    #[test_case(Rule::MissingActionSpecifier, Path::new("C041.f90"))]
    #[test_case(Rule::TrailingBackslash, Path::new("C051.F90"))]
    #[test_case(Rule::MissingIntent, Path::new("C061.f90"))]
    #[test_case(Rule::AssumedSize, Path::new("C071.f90"))]
    #[test_case(Rule::AssumedSizeCharacterIntent, Path::new("C072.f90"))]
    #[test_case(Rule::InitialisationInDeclaration, Path::new("C081.f90"))]
    #[test_case(Rule::PointerInitialisationInDeclaration, Path::new("C082.f90"))]
    #[test_case(Rule::ExternalProcedure, Path::new("C091.f90"))]
    #[test_case(Rule::ProcedureNotInModule, Path::new("C092.f90"))]
    #[test_case(Rule::MissingDefaultPointerInitalisation, Path::new("C101.f90"))]
    #[test_case(Rule::UseAll, Path::new("C121.f90"))]
    #[test_case(Rule::MissingIntrinsic, Path::new("C122.f90"))]
    #[test_case(Rule::MissingAccessibilityStatement, Path::new("C131.f90"))]
    #[test_case(Rule::DefaultPublicAccessibility, Path::new("C132.f90"))]
    #[test_case(Rule::MissingExitOrCycleLabel, Path::new("C141.f90"))]
    #[test_case(Rule::ExitOrCycleInUnlabelledLoop, Path::new("C142.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::ImplicitTyping, Path::new("C001_ok.f90"))]
    #[test_case(Rule::InterfaceImplicitTyping, Path::new("C002_ok.f90"))]
    fn rules_pass(rule_code: Rule, path: &Path) -> Result<()> {
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        assert!(
            diagnostics.is_empty(),
            "Test source has no warnings, but some were raised:\n{diagnostics}"
        );
        Ok(())
    }

    #[test]
    fn warn_c142_on_nested_loops_only() -> Result<()> {
        let rule_code = Rule::ExitOrCycleInUnlabelledLoop;
        let path = Path::new("C142.f90");
        let snapshot = format!(
            "{}_{}_nested_loops_only",
            rule_code.as_ref(),
            path.to_string_lossy()
        );
        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                exit_unlabelled_loops: exit_labels::settings::Settings {
                    allow_unnested_loops: true,
                },
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
