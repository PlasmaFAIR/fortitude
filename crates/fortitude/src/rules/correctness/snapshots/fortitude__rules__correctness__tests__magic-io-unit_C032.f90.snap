---
source: fortitude/src/rules/correctness/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/correctness/C032.f90:4:8: C032 Magic unit '10' in IO statement
  |
2 |   implicit none
3 |   integer :: i, named_unit
4 |   open(10, file="test.txt", action="read")
  |        ^^ C032
5 |   read(10, *) i
6 |   close(10)
  |
  = help: Replace with named variable

./resources/test/fixtures/correctness/C032.f90:5:8: C032 Magic unit '10' in IO statement
  |
3 |   integer :: i, named_unit
4 |   open(10, file="test.txt", action="read")
5 |   read(10, *) i
  |        ^^ C032
6 |   close(10)
  |
  = help: Replace with named variable

./resources/test/fixtures/correctness/C032.f90:6:9: C032 Magic unit '10' in IO statement
  |
4 |   open(10, file="test.txt", action="read")
5 |   read(10, *) i
6 |   close(10)
  |         ^^ C032
7 |
8 |   open(file="test_out.txt", action="write", unit=24)
  |
  = help: Replace with named variable

./resources/test/fixtures/correctness/C032.f90:8:50: C032 Magic unit '24' in IO statement
   |
 6 |   close(10)
 7 |
 8 |   open(file="test_out.txt", action="write", unit=24)
   |                                                  ^^ C032
 9 |   write (fmt=*, unit=24) i
10 |   close(24)
   |
   = help: Replace with named variable

./resources/test/fixtures/correctness/C032.f90:9:22: C032 Magic unit '24' in IO statement
   |
 8 |   open(file="test_out.txt", action="write", unit=24)
 9 |   write (fmt=*, unit=24) i
   |                      ^^ C032
10 |   close(24)
   |
   = help: Replace with named variable

./resources/test/fixtures/correctness/C032.f90:10:9: C032 Magic unit '24' in IO statement
   |
 8 |   open(file="test_out.txt", action="write", unit=24)
 9 |   write (fmt=*, unit=24) i
10 |   close(24)
   |         ^^ C032
11 |
12 |   open(newunit=named_unit, file="test.txt", action="write")
   |
   = help: Replace with named variable
